llext: additional CMake harmonizations

pillo79 opened the pull request:
This PR reworks `add_llext_target` to more cleanly mimic existing CMake machinery, and support adding custom commands to the extension building stages (as turned out to be useful in [this SOF PR](https://github.com/thesofproject/sof/pull/8180)).

The `C_FLAGS` argument to `add_llext_target` has been removed. Flags, defines, paths and similar configuration options can now be set using functions that mimic CMake's `target_`* API.

Custom build steps can be requested via the new `add_llext_command` function, that shares the API with `add_custom_command`:
```CMake
add_llext_command(
   TARGET llext_target
   PRE_BUILD | POST_BUILD | POST_PKG
   COMMAND ...
)
```

Intermediate file or target names are exported via properties on the main target and can be accessed via the `get_target_property()` function:
    - `lib_target`: Target name for the compilation and/or link step.
    - `lib_output`: Name of the binary file resulting from compilation and/or linking steps.
    - `pkg_input`: Name of the binary file to be used as input to the packaging stage.
    - `pkg_output`: Name of the final llext file.
    
The phase options have natural meanings:
* `PRE_BUILD` runs after compilation, but before any linking step takes place (as CMake); might use `lib_target` to get additional variables.
* `POST_BUILD`runs after all code-related compile/link stages have been executed (as CMake), but before packaging; this is expected to create `pkg_input` by processing the contents of `lib_output`.
* `POST_PKG` is a new stage after the packaging in .llext is complete; might access the contents of the file `pkg_output`.

pillo79 committed:
llext: use CMake shared library support on Xtensa

This change reworks the Xtensa support in llext to use CMake's native
shared library support, instead of manually running "gcc -shared".

This change minimizes the differences in llext handling by defining
appropriate CMake targets for the different architectures.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

pillo79 commented:
@lyakh I believe you could use something like this:

```
add_llext_target(hello_world
  OUTPUT  ${llext_bin_file}
  SOURCES ${llext_src_file}
  POSTPROC_CMD  echo
  POSTPROC_ARGS ${CMAKE_C_COMPILER} --
                $<TARGET_PROPERTY:hello_world_lib,COMPILE_FLAGS>
                $<TARGET_PROPERTY:hello_world_lib,LINK_FLAGS>
                $<TARGET_PROPERTY:hello_world_lib,LINK_OPTIONS>
                $<TARGET_OBJECTS:hello_world_lib>
)
```

which results in the following command being run:
```
echo zephyr/twister-out/qemu_xtensa/tests/subsys/llext/llext.simple.xtensa/hello_world/libhello_world_lib.so hello_world_pp.llext zephyr-sdk-0.16.4/xtensa-dc233c_zephyr-elf/bin/xtensa-dc233c_zephyr-elf-gcc --   -gdwarf-4;-fPIC;-nostdlib;-nodefaultlibs zephyr/twister-out/qemu_xtensa/tests/subsys/llext/llext.simple.xtensa/hello_world/CMakeFiles/hello_world_lib.dir/./hello_world.c.obj
```

pillo79 commented:
@tejlmand feedback welcome on [this](https://github.com/zephyrproject-rtos/zephyr/pull/67997/files#diff-099f8d126fd2a4d31179861b2819d9f607cadd14602968912448d7e5d8248b6cR4903-R4904) - these lines are a global CMake config but need to be run _after_ `project()` is executed as far as I understood. Having it in `cmake/modules` or `cmake/compiler` means it is executed before and gets erased. Is this location acceptable?

pillo79 commented:
> @pillo79 some list-conversion missing?

Thanks for noticing, indeed missed a `COMMAND_EXPAND_LISTS`. Fixed!

tejlmand commented on cmake/modules/extensions.cmake:
```diff
@@ -4802,6 +4802,14 @@ endmacro()
+#                    OUTPUT  <output_file>
+#                    SOURCES <source_file>
+#                    [C_FLAGS <flags> ...]
+#                    [POSTPROC_CMD <cmd> [POSTPROC_ARGS <args> ...]]
```
not fond of this, let's split the functions.

`add_llext_target()` creates the target and accepts source files, similar to CMake's add_custom_target / add_library.

But we should consider which one `add_llext_target()` resembles. With this PR I'm starting to think it probably more resembles add_library than custom_target.

Therefore I believe that instead of having `POSTPROC_CMD` then we should be have a function which resembles `add_custom_command(TARGET <target> PRE_/POST_BUILD)`.

As I see, we need finer granularity, so an idea could be:
```
add_llext_command(TARGET <llext_target> 
    [ PRE_BUILD | PRE_STRIP | POST_BUILD | <other> ] # Choose what is appropriate in this context
    COMMAND ....
)
```

That will also allow to gradually build the target and extra steps based on various other setting and not having to know everything from start.

Like:
```
add_llext_target(...)

if(<something>)
    add_llext_command(...)
endif()
```

Note: should we consider to rename `add_llext_target()` to `add_llext_library()` to make it more clear that we actually produce a lib.

teburd replied:
The product is an llext, not necessarily a library, it could be a partially linked elf representing a whole application to load, so I think library is a misrepresentation in this case. If shorter and closer to add _library is needed, then abbreviating to add_llext is fine, its a target to build

tejlmand replied:
> The product is an llext, not necessarily a library, it could be a partially linked elf representing a whole application to load, so I think adding the term library is a misrepresentation in this case.

and yet the term `lib` is used all over the place in this PR:
https://github.com/zephyrproject-rtos/zephyr/blob/37f60ed82af5c0ce18d0ad4c528724900b39f5d9/cmake/modules/extensions.cmake#L5172
https://github.com/zephyrproject-rtos/zephyr/blob/37f60ed82af5c0ce18d0ad4c528724900b39f5d9/cmake/modules/extensions.cmake#L5177
https://github.com/zephyrproject-rtos/zephyr/blob/37f60ed82af5c0ce18d0ad4c528724900b39f5d9/cmake/modules/extensions.cmake#L5195
etc.

tejlmand commented on cmake/modules/extensions.cmake:
```diff
@@ -4802,6 +4802,14 @@ endmacro()
+#   add_llext_target(<target_name>
+#                    OUTPUT  <output_file>
+#                    SOURCES <source_file>
+#                    [C_FLAGS <flags> ...]
```
let's have a dedicate function for `C_FLAGS`, perhaps `llext_compile_options(<target> <options>)`

marc-hb reviewed (commented):
For testing convenience and coverage, could you add some dummy post-processing in one of the samples or tests?

lyakh commented:
Hi, I was able to use this for building an SOF LLEXT binary and it worked well! So you can add my "Tested-by" to your PR if that's something commonly done in Zephyr. But I did have to add `C_FLAGS "-std=gnu99"` and `-shared` to linking flags, maybe they should be added automatically for Xtensa

pillo79 commented:
( DNM since I can't have this reviewed by the release window closing )

pillo79 committed:
llext: add llext_* configuration functions

Remove the C_FLAGS argument from add_llext_target() and add a set of
functions to configure the compilation and linking of an llext using
the same API of the target_* functions.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

pillo79 committed:
llext: add support for build time custom commands

This patch adds support for custom commands to be executed during the
build of an llext target. The commands can be executed at different
points in the build process:

 - PRE_BUILD:  Before the llext code is linked.
 - POST_BUILD: After the llext code is built, but before packaging
	       it in an .llext file.
 - POST_PKG:   After the llext file has been created.

Note that PRE_BUILD is not supported for ARM targets, as in that case
object files are used directly and there is no actual linking step.

The commands can be added using the new add_llext_command() function.
An example usage of it, along with some target properties, is added to
the hello_world test case.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

pillo79 commented:
@tejlmand @teburd @lyakh I should have addressed all concerns.

The function added to customize llext build steps is now exactly like `add_custom_command`:
```CMake
add_llext_command(
   TARGET llext_target
   PRE_BUILD | PRE_PACK | POST_BUILD
   COMMAND ...
) 
``` 
* `PRE_BUILD` and `POST_BUILD` have the usual meanings, while the new PRE_PACK stage is after linking but before "converting to llext" (currently only that `strip`, but maybe in the future more processing will be required? signing?).
 * Everything after `COMMAND` is passed verbatim to the actual `add_custom_command`, so it should be as flexible as the CMake function. 
 * Since it is not possible to make custom targets work like an `add_library`/`add_executable`, I created custom target properties that can be read by user code to get the intermediate stages:
    - `LinkTarget`: Target name for the compilation and/or link step.
    - `LinkedFile`: Object file resulting from compilation and/or linking steps.
    - `PostProcTarget`: Target name for the post-processing step.
    - `PostProcFile`: Post-processed file, before the conversion to an llext file.
    - `OutputFile`: Final llext file.

The `C_FLAGS` argument has also been removed. To set flags, paths, defines et al I have added a set of `llext_` functions that mimic the CMake standard `target_` API but work on the correct private target.

pillo79 commented:
@lyakh your use case is also better supported now, since:
* the full command is available for customizations (you can re-add `-i`/`-o`, sorry for that!);
* no need to hardcode internal target names, they are available via generic properties.
 
I would rewrite the above example as:
```CMake
# Generate the llext target
add_llext_target(hello_world
  OUTPUT  ${llext_bin_file}
  SOURCES ${llext_src_file}
)

# Get internal file and target names 
get_target_property(link_target hello_world LinkTarget)
get_target_property(linked_file hello_world LinkedFile)
get_target_property(postproc_file hello_world PostProcFile)

# Add the custom command using the above variables
add_llext_command(
  TARGET hello_world
  PRE_PACK
  COMMAND echo ${linked_file} ${postproc_file} ${CMAKE_C_COMPILER} --
                $<TARGET_PROPERTY:${link_target},COMPILE_FLAGS>
                $<TARGET_PROPERTY:${link_target},LINK_FLAGS>
                $<TARGET_PROPERTY:${link_target},LINK_OPTIONS>
                $<TARGET_OBJECTS:${link_target}>
)
```

pillo79 commented:
Rebased with `main` to use the new `test/subsys/llext` paths and added a tiny usage example in the `llext.simple` test as per @marc-hb suggestion.

marc-hb reviewed (commented):
I think the terminology can be better and it should be at least more consistent.

You can probably fast-track the first 2 commits in a separate PR?
https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#defining-smaller-prs

pillo79 commented:
Quite a comprehensive review @marc-hb! I understand your PoV and I'll go into detail soon, but a few general remarks first:
* llext was started on ARM with object files, then ported to Xtensa using libraries. There are currently major differences in how these work, and thus we need some ugly logic (both in CMake and in the llext code) to support both.
My goal is to merge these two implementations - I have a draft PR to switch both to dynamic libraries I've been working on since December - and this is already a very tiny step in that direction, but it's just currently not possible. 

* In several places I referred to the "linking" step by considering the Xtensa side, which is more articulate, and also could be the common goal. 
I don't think users should be mindful arch-specific implementation details of how an llext is (currently) built when using `llext_`* API. Developers that need information on how to create an llext using third-party tools should look in the [OT: currently nonexistent] subsys documentation, in my opinion. 
 
---

#### Naming:

There are 3 internal steps when creating an llext:
1. Compiling the source file [or, in the future, files] into object files.
2. Linking them [when a dynamic library is used], or copying the object file if no library is used.
3. Patching the output of step 2 to create a loadable llext file.

On static/dynamic libs created via `add_library()`, CMake: 
* Does not support adding anything before step 1; the same effect would be accomplished with a simple external dependency on the library target.
* Between step 1 and step 2, it executes `PRE_BUILD` / `PRE_LINK` commands (it's really [the same stage](https://cmake.org/cmake/help/latest/command/add_custom_command.html#build-events)).
* After step 2, it executes `POST_BUILD` commands. 
* (Step 3 is specific to llexts).

As per [a previous review](https://github.com/zephyrproject-rtos/zephyr/pull/67997#discussion_r1466270985), I have implemented the same semantics:  
* `PRE_BUILD` before step 2 - if there is one: it makes no sense with `add_library(... OBJECT ...)`;
* `PRE_PACK` after compile (and link), but before the packaging of an llext;
* `POST_BUILD` after the llext target file is complete.

Internal and external variables referring to the "postproc" are indeed confusing leftovers of a previous iteration of the PR. However, I tried - but failed - to come up with valid names aligned to your proposed `PRE_LLEXT` root. I therefore propose to use the following names:

| Internal varitable | External property | Description |
| --- | --- | --- | 
| `llext_lib_target` | `LibTarget` | step 2 (compile+link) target name 
| `llext_lib_output` | `LibOutput`| step 2 output file name 
| `llext_pack_target` | `PackTarget` | step 3 (conversion to `.llext`) target name 
| `llext_pack_input` | `PackInput`| step 3 input file name 
| `llext_pack_output` | `PackOutput`| step 3 output file name 

Could this be a workable solution? Alternative proposals welcome!

pillo79 commented on cmake/modules/extensions.cmake:
```diff
@@ -4971,24 +4971,48 @@ function(add_llext_target target_name)
+  elseif(CONFIG_XTENSA)
+
+    # Ensure shared library support is enabled
+    set_property(GLOBAL PROPERTY TARGET_SUPPORTS_SHARED_LIBS true)
```
This is actually writable (I took this from the SOF PR! :sweat_smile:) and required to enable `add_library(... SHARED ...)`. 
Not sure why or how this is disabled on Xtensa. I would really like to move this to the `cmake/` folder, but I need this line to be executed _after_ the `project()` function, so I [previously asked for guidance](https://github.com/zephyrproject-rtos/zephyr/pull/67997#issuecomment-1905886945)...

tejlmand replied:
> This is actually writable

but it's not supposed to be set after the project() call, and certainly not in a function call like this, which may be called from various locations.

If shared library is supported by the target (Zephyr and Zephyr SDK), then we must find out why this value is `False`, and instead ensure it's properly set to `True`.

I will do some investigations as to why this is not set.

Also consider this comment, which should probably be updated, as there are now cases where we build shared libs:
https://github.com/zephyrproject-rtos/zephyr/blob/569f8f0a43dab5525a03ab18ab4bb50f956b9991/cmake/modules/FindTargetTools.cmake#L77-L78
Note, the `BUILD_SHARED_LIBS` only controls whether shared libs are built when `add_library()` is not specifying the type, not whether shared libs are supported.

marc-hb replied:
> I will do some investigations as to why this is not set.

Awesome, thank you.

pillo79 commented on cmake/modules/extensions.cmake:
```diff
@@ -4971,24 +4971,48 @@ function(add_llext_target target_name)
+    set(linked_file $<TARGET_FILE:${link_target}>)
+
+    # Add the llext flags to the linking step as well
+    # (the linker will ignore the compilation-related ones)
```
Sure, will amend the wording.

pillo79 commented on cmake/modules/extensions.cmake:
```diff
@@ -4999,31 +5023,22 @@ function(add_llext_target target_name)
       OUTPUT ${output_file}
-      COMMAND ${CMAKE_COMMAND} -E copy $<TARGET_OBJECTS:${target_name}_lib> ${output_file}
-      DEPENDS ${target_name}_lib $<TARGET_OBJECTS:${target_name}_lib>
+      COMMAND ${CMAKE_COMMAND} -E copy ${linked_file} ${output_file}
```
This is OT - this PR is not changing current functionality or addressing this topic. 
(My solution would be to move to dynamic libs everywhere, but first need to merge the link code).

teburd replied:
It could work with multiple files, but it would require wrapping an ld merge command...

```sh
ld -relocatable in0.o in1.o -o out.o
```

There's no add_library command for a statically relocatable elf today, we have static, dynamic, and object as options. CMake makes this somewhat annoying to deal with actually.

pillo79 commented on cmake/modules/extensions.cmake:
```diff
@@ -4913,22 +4954,26 @@ endmacro()
+# retrieved using the get_target_property() function:
+#
+# - LinkTarget      Target name for the compilation and/or link step.
+# - LinkedFile      The object file resulting from compilation and/or
```
Discussed in the naming below.

pillo79 commented on cmake/modules/extensions.cmake:
```diff
@@ -5088,10 +5102,92 @@ function(add_llext_target target_name)
+# the build.
+#
+# The differrent build steps are:
+# - PRE_BUILD:  Before the llext code is linked.
```
>  it sounds like it's run between compiling and linking.

In fact, that is [exactly what is happening](https://cmake.org/cmake/help/latest/command/add_custom_command.html#build-events), and the reason why the check just below makes it currently unsupported on ARM, where no linking step is done... :slightly_smiling_face:

pillo79 commented on cmake/modules/extensions.cmake:
```diff
@@ -5072,10 +5086,10 @@ function(add_llext_target target_name)
               $<TARGET_PROPERTY:bintools,strip_flag_outfile>${output_file}
               $<TARGET_PROPERTY:bintools,strip_flag_final>
-      DEPENDS ${link_target} ${linked_file}
+      DEPENDS ${postproc_target} ${postproc_file}
```
Discussed in the naming comment below.

pillo79 commented on cmake/modules/extensions.cmake:
```diff
@@ -5088,10 +5102,92 @@ function(add_llext_target target_name)
+# Usage:
+#   add_llext_command(
+#     TARGET <target_name>
+#     PRE_BUILD | PRE_PACK | POST_BUILD
```
Discussed in the naming comment below.

pillo79 commented on cmake/modules/extensions.cmake:
```diff
@@ -5088,10 +5102,92 @@ function(add_llext_target target_name)
+#
+# The differrent build steps are:
+# - PRE_BUILD:  Before the llext code is linked.
+# - PRE_PACK:   After the llext code is linked, but before converting
```
I will amend the wording to what you proposed, but I disagree with the "users must be familiar" part for the reasons in the main comment.

pillo79 commented on cmake/modules/extensions.cmake:
```diff
@@ -5088,10 +5102,92 @@ function(add_llext_target target_name)
+  # Check that the first unparsed argument is the word COMMAND
+  list(GET LLEXT_UNPARSED_ARGUMENTS 0 command_str)
+  if(NOT command_str STREQUAL "COMMAND")
+    message(FATAL_ERROR "add_llext_command: COMMAND argument must be provided")
```
This is actually still argument parsing, sorted by its position in the command arguments. It turns out the "implementation" part is a single function call... :sweat_smile:

marc-hb commented:
> My goal is to merge these two implementations - I have a draft PR to switch both to dynamic libraries I've been working on since December - and this is already a very tiny step in that direction, but it's just currently not possible.

> In several places I referred to the "linking" step by considering the Xtensa side, which is more articulate, and also could be the common goal.

OK, that really helps. Please, please add a couple TODOs in the code stating this longer term goal and making the nature of that object/library difference clearer. Such TODOs would have saved me a lot of time. Temporary inconsistencies are sometimes required and perfectly fine. On the other hand, believing that some temporary solution is part of the grand design is extremely confusing and time-consuming! Don't write a novel for these TODOs but there needs to be at least one short one in each section where this temporary difference appear (or not far from it).


I tried but it's really hard to find a better name than _PACK for the simple reason that it is "just a `strip` for now" without any indication of what else that step could do in the future... Does anyone have any idea what else there could be there?

Given that lack of foresight, "package" sounds good. But can you please abbreviate it to `_PKG` instead? I think it's more common and less ambiguous. I'll admit I was wondering about "packing" data in the C `struct` sense for a while; something which is not totally foreign to ABIs after all!

I would probably rename POST_BUILD to POST_PKG too. At first sight the consistency with CMake is nice but on second thought one could assume POST_BUILD happens immediately after PRE_BUILD.

marc-hb commented:
Nit: please use the green "Start Review" button to reduce the number of notifications: https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#workflow-suggestions-that-help-reviewers

pillo79 commented:
* Removed all references to obsolete steps, so that now the API publicly only refer to:
   * the "lib" step, for building sources (and linking, on targets using dynamic libs);
   * the "pkg" step, with the Xtensa arch-specific strip command (and any future signing/processing).
* Changed the step names to the more natural `PRE_BUILD`, `POST_BUILD` and `POST_PKG`.
* Added more informative comments and fixed the reviewed ones.

Let me know if I missed something.
@tejlmand your requested changes should have been addressed as well, can you please update your review?

marc-hb commented:
No "deep" review yet but I had very quick look and the last revision feels extremely clear and super easy to make sense of. Thank you!

This can't be completely objective since I lost my "first impression" eyes forever but I  think none of my initial confusion would have happened with this last version.

marc-hb commented on cmake/modules/extensions.cmake:
```diff
@@ -4971,62 +4970,84 @@ function(add_llext_target target_name)
   )
 
-  # Arch-specific conversion of the object file to an llext
+  # Arch-specific packaging of the built binary file in an .llext file
```
```suggestion
  # Arch-specific packaging of the built binary file into an .llext file
```

marc-hb reviewed (dismissed):
I didn't scrutinize every single line looking for a bug but for sure the code changes and commit messages all make sense and they are now all crystal clear to me; should be very "low maintenance".

@lyakh could you test this again? The code has changed a fair bit since you last tested it.

@pillo79 I think you could simply squash the second commit ("internal cosmetic cleanups") into the previous and first commit because the second commit tweaks most of the lines of the first commit. I realize the first commit makes some functional change and the second does not. So "in theory" the split is correct. But I think we have an "exception" here:
- the second commit makes it totally impossible to revert the first commit alone
- This code is too new to be cherry-picked anywhere
- following the evolution of the code and reviewing commit per commit is much harder. This is how I found out. I ended up selecting the first two commits in Github and looked at them together: much easier to read.

While functionally and theoretically different,  in practice these first two commits are "Siamese twins"

pillo79 commented:
>  I think you could simply squash the second commit ("internal cosmetic cleanups") into the previous and first commit 
Done. It has been standing there so long I forgot that first commit already did rework the names :sweat_smile:  Thanks for spotting this.

pillo79 commented:
Rebased on latest main to solve merge conflict with #69689 (moved a CMakeLists.txt).

marc-hb commented on cmake/modules/extensions.cmake:
```diff
@@ -5166,62 +5165,84 @@ function(add_llext_target target_name)
+  # This is currently arch-specific since the ARM loader for .llext files
+  # expects object file format, while the Xtensa one uses shared libraries.
+  set(llext_lib_target ${target_name}_llext_lib)
+  if(CONFIG_ARM)
```
> then having (too much) CONFIG_<feature> is a bad sign.

It is not good but "spreading" is IMHO slightly exaggerated and @pillo79 has committed above to fix this as soon as the corresponding C code is fixed - if nothing else for the simple reason that it makes his own maintenance work harder.

The LLEXT paint is still wet and we need more early adopters (like SOF) to starting using LLEXT in anger and provide feedback to make progress. One thing slowing down adoption and progress has been (as often) the lack of a build system, forcing everyone to write their own, different, custom build script. So @pillo79 stepped up and spent an unbelievable amount of time and patience addressing that gap in the CMake area where extremely few people ever volunteer. Blocking the (extension of) this small, documented CMake hack can re-inforce a chicken-and-egg situation where the corresponding C code lacks users and does not get fixed either.

From experience I know very well that "Fix it later" usually means either "never" or "someone else can do it" but I think this case deserves an exception  because:
- LLEXT is still at the (late) prototype stage so there's barely any user base and nothing "stable" to break yet.
- It's not even a new hack, it's already there
- @pillo79 's dedication is a good indication we can trust him to actually fix this later
- we have no one else :-D

tejlmand replied:
> It is not good but "spreading" is IMHO slightly exaggerated

So which better word would you have in mind to describe the fact that the `if(CONFIG_...)` was used in one location before, and now two more locations are added ?
Or should I not point to the fact that such testing is not desired ?

Also, pointing at this could even have @pillo79 taking a second look at the code to see if this could be combined under the existing single `if(CONFIG_...)` test.

Remember, others may even look at this code and think this is an acceptable way, but for such cases it's always good to have review comments to reference and stating, `yes, but his pattern was accepted under specific conditions, see PR xyz`.

> The LLEXT paint is still wet and we need more early adopters (like SOF) to starting using LLEXT in anger and provide feedback to make progress. 

and I fully agree, and also a reason why I very early suggested that `llext` functions were made first class citizen and placed in `extensions.cmake`. But with that we should also require a bit higher level compared to code in a Zephyr module 🙂 

> @pillo79 's dedication is a good indication we can trust him to actually fix this later

and did I anywhere indicate that this was not the case ?

tejlmand replied:
The fact that something is observed in a PR doesn't mean that it cannot be accepted if reasonable explanation / answers are given.

pillo79 replied:
The "spreading" is due to the fact that build stages are now clearly separated, thus there are multiple (but closely packed) locations in which I need arch-specific behavior. I could probably minimize it somehow, but I chose readability over minimum code.

We can and probably will get rid of the arch-specific tags by defining behavior-specific Kconfigs (for example an enum like `CONFIG_LLEXT_TYPE_DYNAMIC`/`OBJECT`). But this is definitely out of scope of this PR, which is _not_ changing the "hack", just moving it around _to make it more explicit_.

tejlmand commented on cmake/modules/extensions.cmake:
```diff
@@ -5275,3 +5292,91 @@ function(add_llext_target target_name)
+    #   - replace processor "file copy" command with a dummy to prevent
+    #     user files from being clobbered;
+    #   - execute user command(s) after the (empty) processor target.
+    set_property(TARGET ${llext_proc_target} PROPERTY ProcessorCmd
+      ${CMAKE_COMMAND} -E true
+    )
```
not fond of substituting commands this way.

Instead of extra target creation and commands, then has it been considered if the copy here:
https://github.com/zephyrproject-rtos/zephyr/blob/21b29c3d9cc2b90b97001e2ba77caa0d196822fa/cmake/modules/extensions.cmake#L5275

simply can be depending upon the fact of whether a post build command has been defined or not.

Something like:
```
    COMMAND ${CMAKE_COMMAND} -E copy $<IF:$<BOOL:$<TARGET_PROPERTY:${llext_proc_target},${a_property_name}>>,${llext_pkg_input},${llext_lib_output}> ${llext_pkg_output}
```

In the above example, if `a_property_name` (which can be set as `LLEXT_POST_BUILD`) is defined, then the intermediate output file `llext_pkg_input` is copied to `llext_pkg_output`.

But if `a_property_name` is not defined, then `llext_lib_output` is copied directly to `llext_pkg_output`.

Thus the intermediate copy here:
https://github.com/zephyrproject-rtos/zephyr/blob/21b29c3d9cc2b90b97001e2ba77caa0d196822fa/cmake/modules/extensions.cmake#L5266 can be avoided, as well as the custom command here:
https://github.com/zephyrproject-rtos/zephyr/blob/21b29c3d9cc2b90b97001e2ba77caa0d196822fa/cmake/modules/extensions.cmake#L5258
can be removed, afaict.

Probably some more cleanup can be done, but that will depend on the exact solution.

tejlmand replied:
one more observation related to this.
When substituting the copy command with `-E true` or the similar `$<IF:$<BOOL...` construct, then it appears that it must be ensured that the custom llext build step produces `llext_pkg_input`, else later step like: `${CMAKE_COMMAND} -E copy ${llext_pkg_input} ${llext_pkg_output} ` will fail.

But I don't see that mentioned in the function description of `add_llext_command()` or being checked.
Feel free to correct me if I've missed some details in code somewhere.

pillo79 replied:
>  it must be ensured that the custom llext build step produces `llext_pkg_input` [...] I don't see that mentioned in the function description or being checked.

Missed this bit - indeed, will absolutely amend the function doc. Not sure how I could check this in code TBH :sweat_smile:

tejlmand reviewed (dismissed):
Thanks for the additional changes.

Some more comments / questions related to the extended PR.

pillo79 reviewed (commented):
Thanks for the helpful comments @tejlmand. Please see my answers below and let me know how to proceed. Really looking forward to seeing this through :slightly_smiling_face:

pillo79 committed:
cmake: conditionally enable shared library support

When CONFIG_LLEXT is enabled, the Zephyr platform needs to enable dynamic
library support. This is done by setting the `TARGET_SUPPORTS_SHARED_LIBS`
property to `TRUE` in the global property scope.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

pillo79 reviewed (commented):
New update addressing (I hope) all open points:
* The global property change is now addressed at its root, by defining a custom Zephyr `SYSTEM_NAME`,  to properly shadow the original dynamic libs override in `Generic.cmake`.
* Reworked `llext_proc_target`'s custom command so its contents and logic is unambiguous.
* Made the `PRE_BUILD` custom step fail on ARM, since it has no linking step and there is no way to properly implement a workaround for it. 
* Additional function comments explaining expected behavior of the custom commands.

@tejlmand please let me know your feedback.

marc-hb commented:
I just noticed by chance that a (probably small) conflict appeared in `tests/subsys/llext/simple/CMakeLists.txt`. Github does unfortunately not notify about new conflicts.

pillo79 commented:
> a (probably small) conflict appeared in `tests/subsys/llext/simple/CMakeLists.txt`. 

Yes, the sample code needs to be included inside the new `if` block. Rebased to latest main to fix this, no functional change.

teburd commented:
@tejlmand any further feedback? There is other work depending on this

teburd commented:
@tejlmand arguably this is an LLEXT set of changes so I've re-assigned to myself, if this is wrong please do correct it

pillo79 commented:
@marc-hb @lyakh one more review required for this to finally end in the merge queue :partying_face:

marc-hb commented on cmake/modules/extensions.cmake:
```diff
@@ -5280,6 +5302,89 @@ function(add_llext_target target_name)
+# the build. The command will be executed at the specified build step and
+# can refer to <target>'s properties for build-specific details.
+#
+# The differrent build steps are:
```
Minor typo in comment. Obviously not blocking. Do not force-push and lose approvals ONLY for this.

marc-hb reviewed (approved):
I don't understand all the CMake subtleties but everything I understand makes sense. Moreover this PR is very clear, well structured and well commented = highly "maintainable".

I also wonder about the relationship with #70793 but if @lyakh can test this and is happy with it then I'm happy too.

lyakh commented:
> I also wonder about the relationship with #70793 but if @lyakh can test this and is happy with it then I'm happy too.

@marc-hb It is my goal and wish to first establish a PR CI chain for both Zephyr twister and SOF LLEXT use-case. After that I'll gladly continue with my attempt to move over SOF LLEXT cmake code to this new base. I tried it before and it worked well, but we have changed a couple of things since then, so, need to try again. Will work with @pillo79 if any new requirements arise. Yes, I definitely agree that switching over to using these functions would be a good improvement!
